### ISSUE TITLE
False error

### ISSUE THREAD

**Issue by BlackAsLight** [microsoft/TypeScript#51535]
# Bug Report

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBMAWBDAThGBeGByAEgUwBsCQYB6UmAYUQGs8AdMABRCkShCwDoIAHAgJZQAFFiwBKAFAB3eAIJ4YwhCghcFYAOZR44mAG9JMYzAEAzJStQBtAAwBdDOkxZSEg0ZNfZ8xcqQ2DjAAhM7YjFgwAGRRcAFqGtq6Hl6pJlZqEHJmIlJpxgC+nmkZPNm5xTBFXqCQIArqIJr+qmXmuVIFQA)

### 💻 Code

```ts
if (chars[0] === '/')
    while (chars[0] !== '\n' && chars.length)
        chars.shift()
```

### 🙁 Actual behaviour

It throws an error saying there is no overlap between `chars[0]` and `\n` because it believes `chars[0]` is always a `/` and doesn't acknowledge that `chars.shift()` is changing what `char[0]` will be.

### 🙂 Expected behaviour

No error.

**Comment by MartinJohns** [comment]
Duplicate of #9998.

**Comment by fatcerberus** [comment]
okay who gave the grinch a copy of typescript, that can't possibly end well
